README.md
# Genisys connector for Docker

Allows the genisys component to communicate with Docker.

Can manage services as containers.

## Setup

Install the requirements (assuming you got a Python >= 3.4 devel environment ready):

````
$ pip install -r requirements.txt
````

## Run

Start the adapter:

````
$ python genisysconnector.py
````

## Service definition

In order to manage Docker containers for a service, the adapter provides a simple service definition format
to declare how to manage containers associated with a service.

A service definition looks like:

````
myService = {
	"name": "myService"
	"image": "docker_image:tag",
  "command": "command",
	"environment": {
		"ENV_VARIABLE_A": "value",
		"ENV_VARIABLE_B": "value",
	},
	"ports": ['8080']
}
````

A service definition must include a *name* and an *image*, it may optionally provide a *command*, an *environment* hash and an array of *ports*.

The *image* field is used to start a container, if the *command* field has been specified the container will be started using that command.
The *image* field is also used when stopping containers, the image is used as reference to search in running containers.

The *environment* field is used to inject environment variables into the containers associated with the service.

The *ports* field is used to expose a list of ports on the Docker host.

You can specify an optional file called *services.py* at the root of the project and use it to define services using the format defined above.

## HTTP API

The connector exposes a HTTP API. It can be used to perform CRUD actions on services and also to trigger remote procedure calls to manage containers.

Note: The examples use the httpie CLI to query the API, see: https://github.com/jakubroztocil/httpie

### Service HTTP endpoint

The following endpoints are exposed:

* [/service](#service-1) : List service definitions or register a new service definition
* [/service/\<service_name\>](#serviceservice_name) : Retrieve or update a service definition
* [/service/\<service_name\>/scale](#serviceservice_namescale) : Ensure a number of containers are running for a service
* [/service/\<service_name\>/status](#serviceservice_namestatus) : Return the number of running resources for a service

#### /service

This endpoint is used to list service definitions or to create a new service definition.

It supports the following methods: POST and GET.

When hitting the endpoint with a GET, it returns a JSON body like this:

````
{
	myServiceA = {
		"name": "myServiceA"
		"image": "docker_image:tag",
	  "command": "command",
		"environment": {
			"ENV_VARIABLE_A": "value",
			"ENV_VARIABLE_B": "value",
		},
		"ports": ['8080']
	},
	myServiceB = {
		"name": "myServiceB"
		"image": "docker_image:tag",
		"ports": ['5000', '5001'],
		"command": null,
		"environment": null,
	}
}
````

When hitting the endpoint with a POST, it expects a JSON request body that must look like:

````
{
	"name": "service_name",
	"image": "docker_image:tag",
	"command": "command",
	"environment": {
			"ENV_VARIABLE_A": "value",
			"ENV_VARIABLE_B": "value",
		},
		"ports": ['port_number']
	}
}
````

The *name* and *image* fields are mandatory.

The *name* field is used to identify the service.

The *image* field specifies the reference of the container image used when creating/stopping containers. The image tag must be included.

The *command* field specifies which command should be used when starting a container.

The *environment* field is used to inject environment variables into the containers associated with the service.

The *ports* field is used to expose a list of ports on the Docker host.

Example:

````
$ http POST :7051/service name="helloworld" image="tutum/hello-world:latest" ports:='["8080", "8081"]' environment:='{"VAR_A":"value", "VAR_B": "value"}'
````

#### /service/\<service_name\>

This endpoint is used to retrieve a service definition or to update it.

It supports the following methods: PUT and GET.

When hitting the endpoint with a GET, it returns a JSON body like this:

````
{
    "image": "tutum/hello-world:latest",
    "name": "helloworld"
		"command": null,
		"environment": null,
		"ports": null,
}
````

When hitting the endpoint with a PUT, it expects a JSON request body that must look like:

````
{
	"image": "tutum/hello-world:latest",
	"command": "/run.sh",
	"environment": {
		"VAR_A": "value"
	},
	"ports": ['8080'],
}
````

The *image* field is mandatory.

The *image* field specifies the image to use when starting/killing containers. The image tag must be included.

The *command* field specifies which command should be used when starting a container.

The *environment* field is used to inject environment variables into the containers associated with the service.

The *ports* field is used to expose a list of ports on the Docker host.

Example:

````
$ http PUT :7051/service/helloworld image="panamax/hello-world-php:latest" command="/run.sh" ports:='["8080", "8082"]' environment:='{"VAR_A":"value", "VAR_B": "value"}'
````

#### /service/\<service_name\>/scale

This endpoint is used to ensure that a specific number of containers associated to a service are running.

It expects a JSON request body to be POST. The request body must look like:

````
{
	"number": number_of_containers,
}
````

The *number* field is mandatory.

Example:

````
$ http POST :7051/service/helloworld/scale number=3
````

#### /service/\<service_name\>/status

This endpoint returns the number of running resources for a service managed by this connector.

When hitting the endpoint with a GET, it returns a JSON body like this:

````
{
	"running_resources": number_of_running_resources,
}
````


CHANGELOG.md
# Release 0.1.0
The connector is able to manage containers associated to a service.


utils.py
""" Utility module.
Exposes various utility methods.
"""

import errno
from yaml import safe_load


def load_services_from_file(filename):
    """Return service definitions as a dictionary from a file.
    It will read a specific file indicated by a file path, inject its content
    into a dictionary and return it.
    """
    services = {}
    try:
        exec(compile(open(filename, "rb").read(), filename, 'exec'),
             {},
             services)
    except OSError as e:
        if e.errno == errno.ENOENT:
            print("No services definitions file provided.")
        else:
            print("An error occured while trying to read \
                services definitions file. Aborting.")
            raise
    return services


def load_configuration(configuration_file):
    """Return a configuration object from a YAML formatted file."""
    with open(configuration_file, 'r') as stream:
        config = safe_load(stream)
        return config


dockermanager.py
"""Docker management module.
Exposes a class to manage Docker containers using the Python SDK.
"""

from docker import Client, utils


class DockerManager:
    """The DockerManager provides a way to manage a Docker engine."""

    def __init__(self, docker_url):
        """docker_url is the URL to the docker engine. It can either be the URL
        to the docker engine API or a path to the Docker socket.
        """
        self.client = Client(docker_url)

    def count_containers_by_image(self, container_image):
        """Iterates through all running containers to count the number
        of running containers associated with a container image name.
        Returns the number of running container.
        """
        containers = self.client.containers()
        service_container_count = 0
        for container in containers:
            if container['Image'] == container_image:
                service_container_count += 1
        return service_container_count

    def stop_container_by_image(self, container_image):
        """Iterates through all running containers to stop the first
        container that match a container image name.
        """
        containers = self.client.containers()
        for container in containers:
            if container['Image'] == container_image:
                self.client.kill(container['Id'])
                break

    def create_container(self, service_definition):
        """Creates a new container based on a service definition object."""
        container = self.client.create_container(
            image=service_definition['image'],
            command=service_definition['command'],
            ports=service_definition['ports'],
            environment=service_definition['environment'],
            host_config=utils.create_host_config(publish_all_ports=True))
        self.client.start(container.get('Id'))

    def ensure_containers_for_service(self,
                                      service_definition,
                                      container_number):
        """Ensures that a number of container are running for a service.
        Based on the *container_number* parameter it will stop/create
        containers to match this value.
        """
        service_container_count = self.count_containers_by_image(
            service_definition['image'])
        while service_container_count != container_number:
            if service_container_count > container_number:
                self.stop_container_by_image(service_definition['image'])
                service_container_count -= 1
            elif service_container_count < container_number:
                self.create_container(service_definition)
                service_container_count += 1


genisysconnector.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from bottle import post, get, put, run, abort, request
from utils import load_services_from_file
from dockermanager import DockerManager


@get('/service')
def retrieve_service_definitions():
    return services


@post('/service')
def create_service_definition():
    data = request.json
    if not data:
        abort(400, 'No data received')
    try:
        service_name = data["name"]
        services[service_name] = {}
        services[service_name]["name"] = service_name
        services[service_name]["image"] = data["image"]
    except KeyError:
        abort(400, 'Missing parameters.')
    services[service_name]["command"] = data.get("command", None)
    services[service_name]["environment"] = data.get("environment", None)
    services[service_name]["ports"] = data.get("ports", None)


@get('/service/<service_name>')
def retrieve_service_definition(service_name):
    try:
        service_definition = services[service_name]
        return service_definition
    except KeyError:
        abort(501, "Undefined service: %s." % service_name)


@put('/service/<service_name>')
def update_service_definition(service_name):
    data = request.json
    if not data:
        abort(400, 'No data received')
    try:
        services[service_name] = {}
        services[service_name]["name"] = service_name
        services[service_name]["image"] = data["image"]
    except KeyError:
        abort(400, 'Missing parameter.')
    services[service_name]["command"] = data.get("command", None)
    services[service_name]["environment"] = data.get("environment", None)
    services[service_name]["ports"] = data.get("ports", None)


@post('/service/<service_name>/scale')
def scale_service(service_name):
    data = request.json
    if not data:
        abort(400, 'No data received')
    try:
        scale_number = int(data["number"])
    except KeyError:
        abort(400, 'Missing parameter.')
    except ValueError:
        abort(400, 'Invalid value for number parameter.')
    try:
        service_definition = services[service_name]
        docker.ensure_containers_for_service(service_definition, scale_number)
    except KeyError:
        abort(501, "Undefined service: %s." % service_name)


@get('/service/<service_name>/status')
def service_status(service_name):
    try:
        service_definition = services[service_name]
        running_resources = docker.count_containers_by_image(service_definition["image"])
        response = {}
        response["running_resources"] = running_resources
        return response
    except KeyError:
        abort(501, "Undefined service: %s." % service_name)


if __name__ == '__main__':
    services = load_services_from_file("services.py")
    docker = DockerManager('unix://var/run/docker.sock')
    run(host='localhost', port=7051)


main.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

""" Entry point for the connector.
It will start an a web server to expose a HTTP API based on the configuration
defined in the configuration file.
"""

from bottle import post, get, put, run, abort, request
from utils import load_services_from_file, load_configuration
from dockermanager import DockerManager


@get('/service')
def retrieve_service_definitions():
    """Exposes the /service endpoint accessible via a GET method.
    Used to retrieve the service definitions as JSON.
    """
    return services


@post('/service')
def create_service_definition():
    """Exposes the /service endpoint accessible via a POST method.
    Used to create a new service definition.

    It expects a JSON body that looks like:

    {
        "name": "service_name",
        "image": "docker_image:tag",
        "command": "command",
        "environment": {
                "ENV_VARIABLE_A": "value",
                "ENV_VARIABLE_B": "value",
            },
            "ports": ['port_number']
        }
    }

    The *name* and *image* fields are mandatory.

    The *name* field is used to identify the service.

    The *image* field specifies the reference of the container image used
    when creating/stopping containers. The image tag **must** be included.

    The *command* field specifies which command should be used when
    starting a container.

    The *environment* field is used to inject environment variables into
    the containers associated with the service.

    The *ports* field is used to expose a list of ports on the Docker host.
    """
    data = request.json
    if not data:
        abort(400, 'No data received')
    try:
        service_name = data['name']
        services[service_name] = {}
        services[service_name]['name'] = service_name
        services[service_name]['image'] = data['image']
    except KeyError:
        abort(400, 'Missing parameters.')
    services[service_name]['command'] = data.get('command', None)
    services[service_name]['environment'] = data.get('environment', None)
    services[service_name]['ports'] = data.get('ports', None)


@get('/service/<service_name>')
def retrieve_service_definition(service_name):
    """Exposes the /service/<service_name> endpoint accessible via a GET method.
    Used to retrieve the definition of a specific service as JSON.
    """
    try:
        service_definition = services[service_name]
        return service_definition
    except KeyError:
        abort(501, 'Undefined service: %s.' % service_name)


@put('/service/<service_name>')
def update_service_definition(service_name):
    """Exposes the /service/<service_name> endpoint accessible via a PUT method.
    Used to update the definition of an existing service.

    It expects a JSON body that looks like:
    {
        "image": "tutum/hello-world:latest",
        "command": "/run.sh",
        "environment": {
            "VAR_A": "value"
        },
        "ports": ['8080'],
    }
    The image field is mandatory.
    """
    data = request.json
    if not data:
        abort(400, 'No data received')
    try:
        services[service_name] = {}
        services[service_name]['name'] = service_name
        services[service_name]['image'] = data['image']
    except KeyError:
        abort(400, 'Missing parameter.')
    services[service_name]['command'] = data.get('command', None)
    services[service_name]['environment'] = data.get('environment', None)
    services[service_name]['ports'] = data.get('ports', None)


@post('/service/<service_name>/scale')
def scale_service(service_name):
    """Exposes the /service/<service_name>/scale endpoint accessible via
    a POST method.
    Used to scale the number of resources associated to a service.

    It expects a JSON body that looks like:
    {
        "number": number_of_containers,
    }
    """
    data = request.json
    if not data:
        abort(400, 'No data received')
    try:
        scale_number = int(data['number'])
    except KeyError:
        abort(400, 'Missing parameter.')
    except ValueError:
        abort(400, 'Invalid value for number parameter.')
    try:
        service_definition = services[service_name]
        docker.ensure_containers_for_service(service_definition, scale_number)
    except KeyError:
        abort(501, 'Undefined service: %s.' % service_name)


@get('/service/<service_name>/status')
def service_status(service_name):
    """Exposes the /service/<service_name>/status endpoint accessible via
    a GET method.
    Used to retrieve the number of running resources for a service as JSON.
    """
    try:
        service_definition = services[service_name]
        running_resources = docker.count_containers_by_image(
            service_definition['image'])
        response = {}
        response['running_resources'] = running_resources
        return response
    except KeyError:
        abort(501, 'Undefined service: %s.' % service_name)


if __name__ == '__main__':
    """Main method.
    Retrieve the configuration from the configuration file.
    It also retrieves the services from a service definition file.
    Finally, it creates a DockerManager instance and start the web server.
    """
    config = load_configuration('genisys-connector.yml')
    services = load_services_from_file(config['connector']['service_file'])
    docker = DockerManager(config['docker']['url'])
    run(host=config['connector']['bind'],
        port=config['connector']['port'])
